File: src/algorithm7/hcover.py
"""H-cover construction for head-driven parsing.

Implements Definitions 4 and 5 from Satta & Stock (1994).

An h-cover transforms a CFG G into G_H with:
- H-items as nonterminals
- Projection productions P_H^(1): I_D or I_r^(τ-1,τ) → X_H
- Expansion productions P_H^(2): partial item extensions
"""

from dataclasses import dataclass
from typing import Union

from .grammar import Grammar


@dataclass(frozen=True)
class PartialItem:
    """A partial h-item I_r^(s,t).

    Represents partial recognition of production r,
    having processed symbols from boundary s to t (0-indexed, inclusive),
    where 0 ≤ s < τ_r ≤ t ≤ n_r and (s, t) ≠ (0, n_r).
    """
    prod_idx: int  # Production index r
    s: int         # Left boundary (0-indexed)
    t: int         # Right boundary (0-indexed)

    def __repr__(self) -> str:
        return f"I_{self.prod_idx}^({self.s},{self.t})"


@dataclass(frozen=True)
class CompleteItem:
    """A complete h-item I_A.

    Represents complete recognition of nonterminal A.
    """
    symbol: str  # Nonterminal A

    def __repr__(self) -> str:
        return f"I_{self.symbol}"


# Type alias for any h-item
HItem = Union[PartialItem, CompleteItem]


@dataclass(frozen=True)
class ProjectionProd:
    """A projection production from P_H^(1).

    Form: I_D → X_H
    where D is a nonterminal and X is its head.
    """
    lhs: HItem         # I_D (n_r=1) or I_r^(τ-1,τ) (n_r>1)
    head: str          # X_H (the head symbol)
    prod_idx: int      # Which production this projects from

    def __repr__(self) -> str:
        return f"{self.lhs} → {self.head}_H (prod {self.prod_idx})"


@dataclass(frozen=True)
class LeftExpandProd:
    """A left-expansion production from P_H^(2).

    Form: I_r^(s-1,t) → X_{s-1} I_r^(s,t)
    Extends a partial item leftward.
    """
    result: HItem        # I_r^(s-1,t) or I_D (completion)
    symbol: str | None   # X_{s-1} (None means ε-skip)
    source: PartialItem  # I_r^(s,t)
    direction: str = "left"

    def __repr__(self) -> str:
        sym = self.symbol if self.symbol is not None else "ε"
        return f"{self.result} → {sym} {self.source}"


@dataclass(frozen=True)
class RightExpandProd:
    """A right-expansion production from P_H^(2).

    Form: I_r^(s,t+1) → I_r^(s,t) X_{t+1}
    Extends a partial item rightward.
    """
    result: HItem        # I_r^(s,t+1) or I_D (completion)
    source: PartialItem  # I_r^(s,t)
    symbol: str | None   # X_{t+1} (None means ε-skip)
    direction: str = "right"

    def __repr__(self) -> str:
        sym = self.symbol if self.symbol is not None else "ε"
        return f"{self.result} → {self.source} {sym}"


class HCover:
    """The h-cover G_H of a grammar G.

    Provides methods to:
    - Look up projection productions by head symbol
    - Look up expansion productions for extending partial items
    - Look up completion productions for full spans
    """

    def __init__(self, grammar: Grammar):
        self.grammar = grammar
        self.nullables = grammar.nullable_nonterminals()

        # P_H^(1): Projection productions indexed by head symbol
        self.projections: dict[str, list[ProjectionProd]] = {}

        # P_H^(2): Expansion productions indexed by source item
        self.left_expansions: dict[PartialItem, list[LeftExpandProd]] = {}
        self.right_expansions: dict[PartialItem, list[RightExpandProd]] = {}

        self._build()

    def _build(self):
        """Build the h-cover from the grammar."""
        for r, prod in enumerate(self.grammar.productions):
            if len(prod.rhs) == 0:
                continue
            tau = prod.head_pos  # Head position (1-indexed)
            n = len(prod.rhs)    # RHS length

            # P_H^(1): Projection production
            # I_{prod.lhs} → prod.head_H
            head = prod.head
            if n == 1:
                proj_lhs: HItem = CompleteItem(prod.lhs)
            else:
                proj_lhs = PartialItem(r, tau - 1, tau)
            proj = ProjectionProd(lhs=proj_lhs, head=head, prod_idx=r)
            if head not in self.projections:
                self.projections[head] = []
            self.projections[head].append(proj)

            # P_H^(2): Expansions (only for length > 1)
            if n == 1:
                continue

            # Left expansions: I_r^(s,t) -> X_H I_r^(s+1,t)
            for s in range(0, tau - 1):
                for t in range(tau, n + 1):
                    result = PartialItem(r, s, t)
                    if result.s == 0 and result.t == n:
                        continue
                    source = PartialItem(r, s + 1, t)
                    symbol = prod.rhs[s]  # Z_{r,s+1}
                    exp = LeftExpandProd(result, symbol, source)
                    self.left_expansions.setdefault(source, []).append(exp)
                    if symbol in self.nullables:
                        self.left_expansions[source].append(
                            LeftExpandProd(result, None, source)
                        )

            # Right expansions: I_r^(s,t) -> I_r^(s,t-1) Y_H
            for s in range(0, tau):
                for t in range(tau + 1, n + 1):
                    result = PartialItem(r, s, t)
                    if result.s == 0 and result.t == n:
                        continue
                    source = PartialItem(r, s, t - 1)
                    symbol = prod.rhs[t - 1]  # Z_{r,t}
                    exp = RightExpandProd(result, source, symbol)
                    self.right_expansions.setdefault(source, []).append(exp)
                    if symbol in self.nullables:
                        self.right_expansions[source].append(
                            RightExpandProd(result, source, None)
                        )

            # Completions (P_H^(2)(b)) as expansions to complete I_D
            # Left completion: I_D -> X_H I_r^(1,n)
            if tau > 1:
                source = PartialItem(r, 1, n)
                symbol = prod.rhs[0]
                comp = LeftExpandProd(CompleteItem(prod.lhs), symbol, source)
                self.left_expansions.setdefault(source, []).append(comp)
                if symbol in self.nullables:
                    self.left_expansions[source].append(
                        LeftExpandProd(CompleteItem(prod.lhs), None, source)
                    )

            # Right completion: I_D -> I_r^(0,n-1) Y_H
            if tau < n:
                source = PartialItem(r, 0, n - 1)
                symbol = prod.rhs[n - 1]
                comp = RightExpandProd(CompleteItem(prod.lhs), source, symbol)
                self.right_expansions.setdefault(source, []).append(comp)
                if symbol in self.nullables:
                    self.right_expansions[source].append(
                        RightExpandProd(CompleteItem(prod.lhs), source, None)
                    )

    def get_projections(self, head: str) -> list[ProjectionProd]:
        """Get projection productions for a head symbol."""
        return self.projections.get(head, [])

    def get_left_expansions(self, item: PartialItem) -> list[LeftExpandProd]:
        """Get left-expansion productions for a partial item."""
        return self.left_expansions.get(item, [])

    def get_right_expansions(self, item: PartialItem) -> list[RightExpandProd]:
        """Get right-expansion productions for a partial item."""
        return self.right_expansions.get(item, [])

    def initial_item(self, prod_idx: int) -> PartialItem:
        """Get the initial partial item for a production (just the head)."""
        prod = self.grammar.productions[prod_idx]
        tau = prod.head_pos
        if len(prod.rhs) == 1:
            raise ValueError("No partial items for length-1 productions")
        return PartialItem(prod_idx, tau - 1, tau)

    def __repr__(self) -> str:
        return self.format(explain=False)

    def format(self, explain: bool = False) -> str:
        """Format the H-cover for display.

        Args:
            explain: If True, include detailed explanations and source productions
        """
        lines = []

        if explain:
            lines.append("H-Cover G_H (covering grammar for G)")
            lines.append("=" * 50)
            lines.append("")
            lines.append("Note: Boundary indices (s,t) follow Satta & Stock 1994")
            lines.append("  I_r^(s,t) with 0 ≤ s < τ ≤ t ≤ n and (s,t) ≠ (0,n)")
            lines.append("  I_A = complete recognition of nonterminal A")
            lines.append("")
        else:
            lines.append("HCover:")

        # Projections
        if explain:
            lines.append("Projections P_H^(1): I_D -> X_H")
            lines.append("  (To recognize D, first find its head X)")
            lines.append("-" * 40)
        else:
            lines.append("  Projections:")

        for head, projs in self.projections.items():
            for p in projs:
                if explain:
                    prod = self.grammar.productions[p.prod_idx]
                    lines.append(f"  {p.lhs} -> {p.head}_H")
                    lines.append(f"    Source: [{p.prod_idx}] {prod}")
                    lines.append(f"    Meaning: To recognize {p.lhs}, first find head '{p.head}'")
                    lines.append("")
                else:
                    lines.append(f"    {p}")

        # Left Expansions
        if explain:
            lines.append("Left Expansions P_H^(2): I_r^(s-1,t) -> X_{s-1} I_r^(s,t)")
            lines.append("  (Extend partial item leftward by consuming X)")
            lines.append("-" * 40)
        else:
            lines.append("  Left Expansions:")

        for source, exps in self.left_expansions.items():
            for e in exps:
                if explain:
                    prod = self.grammar.productions[e.source.prod_idx]
                    sym = e.symbol if e.symbol is not None else "ε"
                    lines.append(f"  {e.result} -> {sym} {e.source}")
                    lines.append(f"    Source: [{e.source.prod_idx}] {prod}")
                    if isinstance(e.result, PartialItem):
                        lines.append(f"    Meaning: Extend left by consuming '{sym}' at boundary {e.result.s}")
                    else:
                        lines.append(f"    Meaning: Complete {e.result.symbol} by consuming '{sym}' on the left")
                    lines.append("")
                else:
                    lines.append(f"    {e}")

        # Right Expansions
        if explain:
            lines.append("Right Expansions P_H^(2): I_r^(s,t+1) -> I_r^(s,t) X_{t+1}")
            lines.append("  (Extend partial item rightward by consuming X)")
            lines.append("-" * 40)
        else:
            lines.append("  Right Expansions:")

        for source, exps in self.right_expansions.items():
            for e in exps:
                if explain:
                    prod = self.grammar.productions[e.source.prod_idx]
                    sym = e.symbol if e.symbol is not None else "ε"
                    lines.append(f"  {e.result} -> {e.source} {sym}")
                    lines.append(f"    Source: [{e.source.prod_idx}] {prod}")
                    if isinstance(e.result, PartialItem):
                        lines.append(f"    Meaning: Extend right by consuming '{sym}' at boundary {e.result.t}")
                    else:
                        lines.append(f"    Meaning: Complete {e.result.symbol} by consuming '{sym}' on the right")
                    lines.append("")
                else:
                    lines.append(f"    {e}")

        return "\n".join(lines)

File: ocaml/lib/hcover.ml
(** H-cover construction for head-driven parsing. *)

module StringSet = Set.Make(String)

type partial_item = {
  prod_idx : int;
  s : int;
  t : int;
}

type complete_item = {
  symbol : Grammar.symbol;
}

type hitem =
  | Partial of partial_item
  | Complete of complete_item

type projection_prod = {
  lhs : hitem;
  head : Grammar.symbol;
  proj_prod_idx : int;
}

type left_expand_prod = {
  left_result : hitem;
  left_symbol : Grammar.symbol option;
  left_source : partial_item;
}

type right_expand_prod = {
  right_result : hitem;
  right_source : partial_item;
  right_symbol : Grammar.symbol option;
}

(* Comparison functions *)
let compare_partial a b =
  let c = Int.compare a.prod_idx b.prod_idx in
  if c <> 0 then c else
  let c = Int.compare a.s b.s in
  if c <> 0 then c else
  Int.compare a.t b.t

let compare_complete a b = String.compare a.symbol b.symbol

let compare_hitem a b =
  match a, b with
  | Partial pa, Partial pb -> compare_partial pa pb
  | Complete ca, Complete cb -> compare_complete ca cb
  | Partial _, Complete _ -> -1
  | Complete _, Partial _ -> 1

(* Hash functions for Hashtbl keys *)
let hash_partial p =
  Hashtbl.hash (p.prod_idx, p.s, p.t)

module PartialItemKey = struct
  type t = partial_item
  let equal a b = compare_partial a b = 0
  let hash = hash_partial
end

module PartialTbl = Hashtbl.Make(PartialItemKey)

type t = {
  grammar : Grammar.t;
  projections : (Grammar.symbol, projection_prod list) Hashtbl.t;
  left_expansions : left_expand_prod list PartialTbl.t;
  right_expansions : right_expand_prod list PartialTbl.t;
}

let grammar t = t.grammar

let build grammar =
  let projections = Hashtbl.create 16 in
  let left_expansions = PartialTbl.create 64 in
  let right_expansions = PartialTbl.create 64 in

  let nullable =
    Grammar.nullable_nonterminals grammar
    |> List.fold_left (fun acc sym -> StringSet.add sym acc) StringSet.empty
  in
  let is_nullable sym = StringSet.mem sym nullable in

  let prods = Grammar.productions grammar in
  Array.iteri (fun r prod ->
    let tau = prod.Grammar.head_pos in
    let n = Array.length prod.Grammar.rhs in
    if n = 0 then
      ()
    else begin
    let head = Grammar.head prod in

    (* Projection production: I_{prod.lhs} -> head_H *)
    let proj_lhs =
      if n = 1 then
        Complete { symbol = prod.lhs }
      else
        Partial { prod_idx = r; s = tau - 1; t = tau }
    in
    let proj = { lhs = proj_lhs; head; proj_prod_idx = r } in
    let existing = Hashtbl.find_opt projections head |> Option.value ~default:[] in
    Hashtbl.replace projections head (proj :: existing);

    (* Expansions for length > 1 *)
    if n > 1 then begin
      (* Left expansions: I_r^(s,t) -> X_H I_r^(s+1,t) *)
      for s = 0 to tau - 2 do
        for t = tau to n do
          let result = { prod_idx = r; s; t } in
          if not (result.s = 0 && result.t = n) then begin
            let source = { prod_idx = r; s = s + 1; t } in
            let symbol = prod.Grammar.rhs.(s) in
            let exp = { left_result = Partial result; left_symbol = Some symbol; left_source = source } in
            let existing = PartialTbl.find_opt left_expansions source |> Option.value ~default:[] in
            let exps = exp :: existing in
            let exps =
              if is_nullable symbol then
                { left_result = Partial result; left_symbol = None; left_source = source } :: exps
              else
                exps
            in
            PartialTbl.replace left_expansions source exps
          end
        done
      done;

      (* Right expansions: I_r^(s,t) -> I_r^(s,t-1) Y_H *)
      for s = 0 to tau - 1 do
        for t = tau + 1 to n do
          let result = { prod_idx = r; s; t } in
          if not (result.s = 0 && result.t = n) then begin
            let source = { prod_idx = r; s; t = t - 1 } in
            let symbol = prod.Grammar.rhs.(t - 1) in
            let exp = { right_result = Partial result; right_source = source; right_symbol = Some symbol } in
            let existing = PartialTbl.find_opt right_expansions source |> Option.value ~default:[] in
            let exps = exp :: existing in
            let exps =
              if is_nullable symbol then
                { right_result = Partial result; right_source = source; right_symbol = None } :: exps
              else
                exps
            in
            PartialTbl.replace right_expansions source exps
          end
        done
      done;

      (* Completions as expansions *)
      if tau > 1 then begin
        let source = { prod_idx = r; s = 1; t = n } in
        let symbol = prod.Grammar.rhs.(0) in
        let exp = { left_result = Complete { symbol = prod.lhs }; left_symbol = Some symbol; left_source = source } in
        let existing = PartialTbl.find_opt left_expansions source |> Option.value ~default:[] in
        let exps = exp :: existing in
        let exps =
          if is_nullable symbol then
            { left_result = Complete { symbol = prod.lhs }; left_symbol = None; left_source = source } :: exps
          else
            exps
        in
        PartialTbl.replace left_expansions source exps
      end;

      if tau < n then begin
        let source = { prod_idx = r; s = 0; t = n - 1 } in
        let symbol = prod.Grammar.rhs.(n - 1) in
        let exp = { right_result = Complete { symbol = prod.lhs }; right_source = source; right_symbol = Some symbol } in
        let existing = PartialTbl.find_opt right_expansions source |> Option.value ~default:[] in
        let exps = exp :: existing in
        let exps =
          if is_nullable symbol then
            { right_result = Complete { symbol = prod.lhs }; right_source = source; right_symbol = None } :: exps
          else
            exps
        in
        PartialTbl.replace right_expansions source exps
      end
    end
    end
  ) prods;

  { grammar; projections; left_expansions; right_expansions }

let create = build

let get_projections t head =
  Hashtbl.find_opt t.projections head |> Option.value ~default:[]

let get_left_expansions t item =
  PartialTbl.find_opt t.left_expansions item |> Option.value ~default:[]

let get_right_expansions t item =
  PartialTbl.find_opt t.right_expansions item |> Option.value ~default:[]

let initial_item t prod_idx =
  let prod = Grammar.get_production t.grammar prod_idx in
  let tau = prod.head_pos in
  if Array.length prod.rhs = 1 then
    invalid_arg "No partial items for length-1 productions";
  { prod_idx; s = tau - 1; t = tau }

(* Pretty printers *)
let pp_partial fmt p =
  Format.fprintf fmt "I_%d^(%d,%d)" p.prod_idx p.s p.t

let pp_complete fmt c =
  Format.fprintf fmt "I_%s" c.symbol

let pp_symbol_opt fmt = function
  | None -> Format.fprintf fmt "ε"
  | Some sym -> Format.fprintf fmt "%s" sym

let symbol_opt_to_string = function
  | None -> "ε"
  | Some sym -> sym

let pp_hitem fmt = function
  | Partial p -> pp_partial fmt p
  | Complete c -> pp_complete fmt c

let pp fmt t =
  Format.fprintf fmt "HCover:@.";
  Format.fprintf fmt "  Projections:@.";
  Hashtbl.iter (fun _head projs ->
    List.iter (fun p ->
      Format.fprintf fmt "    %a -> %s_H (prod %d)@."
        pp_hitem p.lhs p.head p.proj_prod_idx
    ) projs
  ) t.projections;
  Format.fprintf fmt "  Left Expansions:@.";
  PartialTbl.iter (fun _source exps ->
    List.iter (fun e ->
      Format.fprintf fmt "    %a -> %a %a@."
        pp_hitem e.left_result pp_symbol_opt e.left_symbol pp_partial e.left_source
    ) exps
  ) t.left_expansions;
  Format.fprintf fmt "  Right Expansions:@.";
  PartialTbl.iter (fun _source exps ->
    List.iter (fun e ->
      Format.fprintf fmt "    %a -> %a %a@."
        pp_hitem e.right_result pp_partial e.right_source pp_symbol_opt e.right_symbol
    ) exps
  ) t.right_expansions

let pp_explain fmt t =
  let prods = Grammar.productions t.grammar in
  Format.fprintf fmt "H-Cover G_H (covering grammar for G)@.";
  Format.fprintf fmt "==================================================@.";
  Format.fprintf fmt "@.";
  Format.fprintf fmt "Note: Boundary indices (s,t) follow Satta & Stock 1994@.";
  Format.fprintf fmt "  I_r^(s,t) with 0 ≤ s < τ ≤ t ≤ n and (s,t) ≠ (0,n)@.";
  Format.fprintf fmt "  I_A = complete recognition of nonterminal A@.";
  Format.fprintf fmt "@.";

  (* Projections *)
  Format.fprintf fmt "Projections P_H^(1): I_D -> X_H@.";
  Format.fprintf fmt "  (To recognize D, first find its head X)@.";
  Format.fprintf fmt "----------------------------------------@.";
  Hashtbl.iter (fun _head projs ->
    List.iter (fun p ->
      let prod = prods.(p.proj_prod_idx) in
      Format.fprintf fmt "  %a -> %s_H@."
        pp_hitem p.lhs p.head;
      Format.fprintf fmt "    Source: [%d] %a@."
        p.proj_prod_idx Grammar.pp_production prod;
      Format.fprintf fmt "    Meaning: To recognize %a, first find head '%s'@."
        pp_hitem p.lhs p.head;
      Format.fprintf fmt "@."
    ) projs
  ) t.projections;

  (* Left Expansions *)
  Format.fprintf fmt "Left Expansions P_H^(2): I_r^(s-1,t) -> X_{s-1} I_r^(s,t)@.";
  Format.fprintf fmt "  (Extend partial item leftward by consuming X)@.";
  Format.fprintf fmt "----------------------------------------@.";
  PartialTbl.iter (fun _source exps ->
    List.iter (fun e ->
      let prod = prods.(e.left_source.prod_idx) in
      Format.fprintf fmt "  %a -> %a %a@."
        pp_hitem e.left_result pp_symbol_opt e.left_symbol pp_partial e.left_source;
      Format.fprintf fmt "    Source: [%d] %a@."
        e.left_source.prod_idx Grammar.pp_production prod;
      (match e.left_result with
       | Partial p ->
         Format.fprintf fmt "    Meaning: Extend left by consuming '%s' at boundary %d@."
           (symbol_opt_to_string e.left_symbol) p.s
       | Complete c ->
         Format.fprintf fmt "    Meaning: Complete %s by consuming '%s' on the left@."
           c.symbol (symbol_opt_to_string e.left_symbol));
      Format.fprintf fmt "@."
    ) exps
  ) t.left_expansions;

  (* Right Expansions *)
  Format.fprintf fmt "Right Expansions P_H^(2): I_r^(s,t+1) -> I_r^(s,t) X_{t+1}@.";
  Format.fprintf fmt "  (Extend partial item rightward by consuming X)@.";
  Format.fprintf fmt "----------------------------------------@.";
  PartialTbl.iter (fun _source exps ->
    List.iter (fun e ->
      let prod = prods.(e.right_source.prod_idx) in
      Format.fprintf fmt "  %a -> %a %a@."
        pp_hitem e.right_result pp_partial e.right_source pp_symbol_opt e.right_symbol;
      Format.fprintf fmt "    Source: [%d] %a@."
        e.right_source.prod_idx Grammar.pp_production prod;
      (match e.right_result with
       | Partial p ->
         Format.fprintf fmt "    Meaning: Extend right by consuming '%s' at boundary %d@."
           (symbol_opt_to_string e.right_symbol) p.t
       | Complete c ->
         Format.fprintf fmt "    Meaning: Complete %s by consuming '%s' on the right@."
           c.symbol (symbol_opt_to_string e.right_symbol));
      Format.fprintf fmt "@."
    ) exps
  ) t.right_expansions
